src/lib.rs

```diff
@@ -1,14 +1,363 @@
-pub fn add(left: u64, right: u64) -> u64 {
-    left + right
+use std::hash::{DefaultHasher, Hash, Hasher};
+
+/// Returns the hash of a single `Hash` value.
+fn hash_single<H: Hash>(value: H) -> u64 {
+    let mut hasher = DefaultHasher::new();
+    value.hash(&mut hasher);
+    hasher.finish()
+}
+
+/// Returns the hash resulting of combining two `Hash` values.
+fn hash_pair<H: Hash>(first: H, second: H) -> u64 {
+    let mut hasher = DefaultHasher::new();
+    first.hash(&mut hasher);
+    second.hash(&mut hasher);
+    hasher.finish()
+}
+
+/// Returns a given index's nth ancestor's index under its respective level.
+/// * `index` - The child index.
+/// * `level` - The relative upwards level of the target ancestor.
+fn ancestor_index(index: usize, level: usize) -> usize {
+    index / (2_usize.pow(level as u32)) // Integer division.
+}
+
+/// Returns the index of the sibling node of the given index.
+/// * `index` - The target index.
+fn sibling_index(index: usize) -> usize {
+    if index % 2 == 0 { index + 1 } else { index - 1 }
+}
+
+/// Given the starting leaves, generates all the upper levels of the tree
+/// by computing the hashes of each pair iteratively.
+/// * `leaves` - Level 0, the starting leaves' hashes.
+/// * `levels` - Vector where the generated levels will be stored.
+fn generate_tree_levels(leaves: &Vec<u64>, levels: &mut Vec<Vec<u64>>) {
+    let mut current: Vec<u64> = leaves.to_owned();
+    levels.push(current.clone());
+
+    while current.len() > 1 {
+        let current_len = current.len();
+        let mut next_level = Vec::new();
+        for index in (0..current_len).step_by(2) {
+            let hash = hash_pair(current[index], current[index + 1]);
+            next_level.push(hash);
+        }
+        current = next_level;
+        levels.push(current.clone());
+    }
+}
+
+/// Base structure were merke tree data is stored.
+pub struct MerkleTree {
+    levels: Vec<Vec<u64>>,
+    capacity: usize,
+    padding: usize,
+}
+
+/// Contains merkle proof information for later validation.
+pub enum MerkleProof {
+    Proof {
+        index: usize,
+        nodes: Vec<u64>,
+        root: u64,
+    },
+    Invalid,
+}
+
+impl MerkleTree {
+    /// The default `Hash` value that is used as padding.
+    fn pad() -> u64 {
+        0
+    }
```

Comment:
Is this necessary to be a function? Can't this be a constant?

Reply:
Agree. I updated it.